### DISCUSSION TITLE
Channel callback from delivery acknowledgement timeout.

### DISCUSSION THREAD

**Discussion by dchompd** [ruby-amqp/bunny#682] [General]
Consumer silently losing its channel due to RabbitMQ closing the channel because of a delivery acknowledgement timeout.

I have a channel open with a consumer working it.  Messages are being being pulled off and the consumer is manually acknowledging them.  I ran into the issue of a delivery acknowledgement timeout (https://www.rabbitmq.com/docs/consumers#acknowledgement-timeout) that resulted in rabbit closing the consumers channel with a `PRECONDITION_FAILED` channel exception.  This manifested in the application silently accumulating messages with no indication that the consumer was no longer being delivered messages on the channel.  When I recreate the delivery ack timeout (create q with a short timeout and simulate a job that hangs the work pool thread and never acks), I see the `AMQ::Protocol::Channel::Close` frame  (https://github.com/ruby-amqp/bunny/blob/8b700a00b62d4f887f53d4871f019d6d494c5f4d/lib/bunny/channel.rb#L1802) and also noticed there is the `@on_error`  handler.  Unfortunately, in this case, the on_error handler is never called (https://github.com/ruby-amqp/bunny/blob/8b700a00b62d4f887f53d4871f019d6d494c5f4d/lib/bunny/channel.rb#L1807-L1812) and the `last_channel_error` ends up pushed onto the continuations.  If the thread working the message is hung, this error is never raised.  I'm trying to understand if there is a way my channel can be notified that RabbitMQ is closing it due to the delivery acknowledgement timeout (reply code 406)? (besides monkey patching a condition into `channel_level_exception_after_operation_that_has_no_response?` https://github.com/ruby-amqp/bunny/blob/8b700a00b62d4f887f53d4871f019d6d494c5f4d/lib/bunny/channel.rb#L1822-L1824).  My current monkey patch just adds an additional text to that line 
```ruby 
method.reply_code == 406 && (method.reply_text =~ /unknown delivery tag/ || method.reply_text =~ /delivery acknowledgement on channel \d+ timed out/)
``` 
but I'm trying to fully understand what the best approach here would be.  Thanks!

**Comment by michaelklishin** [comment]
Consumers do not manage channels. If RabbitMQ closes the channel, Bunny cannot do anything about it (it also works the other way around).

Asynchronous error notifications is the only option available to client libraries: Java and .NET clients expose "shutdown handlers" on their `Consumer` interfaces.

Bunny allows you to [pass in a block](https://github.com/ruby-amqp/bunny/blob/main/lib/bunny/channel.rb#L1546-L1556) for error handling. A similar method can be added to the consumer interface but since most consumers are `subscribe` blocks, I don't have a good idea as to how exactly this might work.

Using exceptions can lead to very difficult to reason about behaviors. In fact, your idea of updating `channel_level_exception_after_operation_that_has_no_response` seems the most pragmatic one, despite depending on a (rarely changing) error message text.

Please submit a PR.

**Comment by dchompd** [comment]
Thank you for the additional information. I opened a PR with the above changes.  PR: https://github.com/ruby-amqp/bunny/pull/684
